File: harness/bunnymark/src/main/kotlin/godot/benchmark/bunnymark/BunnymarkV1Sprites.kt
package godot.benchmark.bunnymark

import godot.Node2D
import godot.RandomNumberGenerator
import godot.ResourceLoader
import godot.Sprite2D
import godot.Texture2D
import godot.annotation.GodotMember
import godot.annotation.GodotScript
import godot.core.Vector2
import godot.core.signal1


@GodotScript("BunnymarkV1Sprites")
class BunnymarkV1Sprites : Node2D() {

	val benchmarkFinished by signal1<Int>("bunnyCount")

	private data class Bunny(var sprite: Sprite2D, var speed: Vector2)

	private val bunnies = mutableListOf<Bunny>()
	private val gravity = 500
	private val bunnyTexture = ResourceLoader.load("res://images/godot_bunny.png") as Texture2D
	private val randomNumberGenerator = RandomNumberGenerator()

	private lateinit var screenSize: Vector2

	override fun _ready() {
		randomNumberGenerator.randomize()
	}

	override fun _process(delta: Double) {
		screenSize = getViewportRect().size

		for (bunny in bunnies) {
			val pos = bunny.sprite.position
			val speed = bunny.speed

			pos.x += speed.x * delta
			pos.y += speed.y * delta

			speed.y += gravity * delta

			if (pos.x > screenSize.x) {
				speed.x *= -1
				pos.x = screenSize.x
			}

			if (pos.x < 0) {
				speed.x *= -1.0
				pos.x = 0.0
			}

			if (pos.y > screenSize.y) {
				pos.y = screenSize.y
				if (randomNumberGenerator.randf() > 0.5) {
					speed.y = -(randomNumberGenerator.randi() % 1100 + 50).toDouble()
				} else {
					speed.y *= -0.85
				}
			}

			if (pos.y < 0) {
				speed.y = 0.0
				pos.y = 0.0
			}

			bunny.sprite.position = pos
			bunny.speed = speed
		}
	}

	@GodotMember
	fun addBunny() {
		val bunny = Sprite2D()
		bunny.texture = bunnyTexture
		addChild(bunny)
		bunny.position = Vector2(screenSize.x / 2, screenSize.y / 2)
		bunnies.add(
			Bunny(
				bunny,
				Vector2(randomNumberGenerator.randi() % 200 + 50, randomNumberGenerator.randi() % 200 + 50)
			)
		)
	}

	@GodotMember
	fun removeBunny() {
		if (bunnies.size == 0) return
		val bunny = bunnies[bunnies.size - 1]
		removeChild(bunny.sprite)
		bunnies.removeAt(bunnies.size - 1)
        bunny.sprite.queueFree()
	}

	@GodotMember
	fun finish() {
        benchmarkFinished.emit(bunnies.size)
	}
}

File: harness/bunnymark/benchmarks/BunnymarkV1Sprites/cs/BunnymarkV1Sprites.cs
using Godot;
using System;
using System.Collections.Generic;

public partial class BunnymarkV1Sprites : Node2D
{
    private class Pair
    {
        public Sprite2D Sprite2D;
        public Vector2 Vector;
    }

    List<Pair> bunnies = new List<Pair>();
    Vector2 screenSize;

    Texture2D bunnyTexture = (Texture2D)GD.Load("res://images/godot_bunny.png");
    Random random = new Random();
    int gravity = 500;

    public override void _Process(float delta)
    {
        screenSize = GetViewportRect().Size;

        foreach (var bunny in bunnies)
        {
            var position = bunny.Sprite2D.Position;
            var newPosition = bunny.Vector;

            position.x += newPosition.x * delta;
            position.y += newPosition.y * delta;

            newPosition.y += gravity * delta;

            if (position.x > screenSize.x)
            {
                newPosition.x *= -1;
                position.x = screenSize.x;
            }

            if (position.x < 0)
            {
                newPosition.x *= -1;
                position.x = 0;
            }

            if (position.y > screenSize.y)
            {
                position.y = screenSize.y;
                if (random.NextDouble() > 0.5)
                {
                    newPosition.y = (random.Next() % 1100 + 50);
                }
                else
                {
                    newPosition.y *= -0.85f;
                }
            }

            if (position.y < 0)
            {
                newPosition.y = 0;
                position.y = 0;
            }

            bunny.Sprite2D.Position = position;
            bunny.Vector = newPosition;
        }
    }

    public void add_bunny()
    {
        var bunny = new Sprite2D();
        bunny.SetTexture(bunnyTexture);
        AddChild(bunny);
        bunny.Position = new Vector2(screenSize.x / 2, screenSize.y / 2);
        bunnies.Add(new Pair() { Sprite2D = bunny, Vector = new Vector2(random.Next() % 200 + 50, random.Next() % 200 + 50) });
    }

    public void remove_bunny()
    {
        if (bunnies.Count == 0) {
			return;
		}

        var bunny = bunnies[bunnies.Count - 1];
        bunnies.RemoveAt(bunnies.Count - 1);
        RemoveChild(bunny.Sprite2D);
        bunny.Sprite2D.QueueFree();
    }

    public void finish()
    {
        EmitSignal("benchmark_finished", bunnies.Count);
    }
}
